pkg/gardenlet/controller/shoot/shoot/cleaner.go

```diff
@@ -0,0 +1,204 @@
+}
+
+// WaitUntilManagedResourcesDeleted waits until all ManagedResource resources in the shoot namespace have been deleted.
+func (c *cleaner) WaitUntilManagedResourcesDeleted(ctx context.Context) error {
+	return kubernetesutils.WaitUntilResourcesDeleted(ctx, c.seedClient, &resourcesv1alpha1.ManagedResourceList{}, defaultInterval, client.InNamespace(c.seedNamespace))
+}
+
+// DeleteSecrets removes all remaining finalizers and deletes all secrets in the shoot namespace.
+func (c *cleaner) DeleteSecrets(ctx context.Context) error {
+	c.log.Info("Deleting all secrets in namespace", "namespace", c.seedNamespace)
+	if err := c.seedClient.DeleteAllOf(ctx, &corev1.Secret{}, client.InNamespace(c.seedNamespace)); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+	return c.removeFinalizersFromObjects(ctx, c.seedNamespace, &corev1.SecretList{})
+}
+
+// DeleteCluster deletes the shoot Cluster resource in the seed cluster.
+func (c *cleaner) DeleteCluster(ctx context.Context) error {
+	cluster := c.getEmptyCluster()
+
+	c.log.Info("Deleting Cluster resource", "cluster", cluster.Name)
+	if err := client.IgnoreNotFound(c.seedClient.Delete(ctx, cluster)); err != nil {
+		return err
+	}
+
+	return controllerutils.RemoveAllFinalizers(ctx, c.seedClient, cluster)
+}
+
+// WaitUntilClusterDeleted waits until the shoot Cluster resource in the seed cluster has been deleted.
+func (c *cleaner) WaitUntilClusterDeleted(ctx context.Context) error {
+	return kubernetesutils.WaitUntilResourceDeleted(ctx, c.seedClient, c.getEmptyCluster(), defaultInterval)
+}
+
+func (c *cleaner) getEmptyCluster() *extensionsv1alpha1.Cluster {
+	return &extensionsv1alpha1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: c.seedNamespace}}
+}
+
+func (c *cleaner) removeFinalizersFromObjects(ctx context.Context, namespace string, objectList client.ObjectList) error {
+	return c.applyToObjects(ctx, namespace, objectList, func(ctx context.Context, object client.Object) error {
+		if len(object.GetFinalizers()) > 0 {
+			c.log.Info("Removing finalizers", "kind", object.GetObjectKind().GroupVersionKind().Kind, "object", client.ObjectKeyFromObject(object))
+			return controllerutils.RemoveAllFinalizers(ctx, c.seedClient, object)
+		}
+		return nil
+	})
+}
+
+func (c *cleaner) applyToObjects(ctx context.Context, namespace string, objectList client.ObjectList, fn func(ctx context.Context, object client.Object) error) error {
+	if err := c.seedClient.List(ctx, objectList, client.InNamespace(namespace)); err != nil {
+		return err
+	}
+	return utilclient.ApplyToObjects(ctx, objectList, fn)
+}
+
+func (c *cleaner) finalizeShootManagedResources(ctx context.Context, namespace string) error {
+	mrList := &resourcesv1alpha1.ManagedResourceList{}
+	if err := c.seedClient.List(ctx, mrList, client.InNamespace(namespace)); err != nil {
+		return err
+	}
+
+	shootMRList := &resourcesv1alpha1.ManagedResourceList{}
+	for _, mr := range mrList.Items {
+		if mr.Spec.Class != nil {
+			continue
+		}
```

Comment:
Please add a similar TODO statement here like this: https://github.com/gardener/gardener/blob/b9651c20c89d4b19a7615873d56f2aa71381e624/pkg/resourcemanager/controller/add.go#L121-L122